### PR TITLE
Add user-configurable overrides module

### PR DESCRIPTION
**Pull request by kvrhdn**
**What this PR does**:
Introduces a new optional feature: user-configurable overrides. Goal is to allow tenants to change a subset of overrides using an API. Changes will be limited to `forwarders` and later several `metrics_generator_...` overrides.

#### Architecture

For every tenant we will store an `overrides.json` file in a bucket. These files will be stored in a separate bucket from the backend.

```
Overrides bucket
├── 1
│   └── overrides.json
└── 2
    └── overrides.json
```

This bucket is regularly polled and a copy of the limits is kept in-memory. When requesting the overrides for a tenant, the overrides module will now:
1. check this override is set in the user-configurable overrides, if so return that value 
2. check if this override is set in the runtime config (configmap), if so return that value
3. return the default values

#### API

```
GET /api/overrides
{
  forwarders: ["k6-cloud-insights-staging"],
}
```

```
PUT /api/overrides
{
  forwarders: ["k6-cloud-insights-staging"],
}
```

```
DELETE /api/overrides
```

Later we might consider adding a PATCH endpoint to perform partial updates/deletes.

**Which issue(s) this PR fixes**:
~~Fixes #<issue number>~~ New feature!

**Checklist**
- [x] Tests updated
- [ ] Documentation added https://github.com/grafana/tempo/issues/2627
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/overrides/user_config_overrides.go
Comment (electron0zero) [inline modules/overrides/user_config_overrides.go]: this will only return configured overrides when configured overrides are set, and runtime overrides when configured overrides are not set.

I am leaning towards keeping it as is. do we want this behaviour and want to merge both Forwarders?

**Comment by knylander-grafana** [comment]
@kvrhdn Will we need doc for this? Should I create a doc issue for it or is it something you'd add to this PR?

**Comment by kvrhdn** [comment]
> @kvrhdn Will we need doc for this? Should I create a doc issue for it or is it something you'd add to this PR?

Yes! This is adding a new feature with an API endpoint and it's own set of configurations. Feel free to create a doc issue and assign it to me, I can get a PR started and we can iterate from there 🙂

File: modules/overrides/http.go
Comment (zalegrala) [inline modules/overrides/http.go:21]: Do you want to set the span error if there is one?

File: modules/overrides/user_config_overrides.go
Comment (zalegrala) [inline modules/overrides/user_config_overrides.go:159]: Consider renaming to `running` for convention.  Not a blocker.

**Review by zalegrala — COMMENTED** [review]
This looks pretty nice and clean to me.  :+1: Happy to approve, but will give others a chance to review also.

File: integration/e2e/overrides_client.go
Comment (joe-elliott) [inline integration/e2e/overrides_client.go:21]: current API client is in `/pkg/util/client.go`. i'm thinking we should move both of these to a new package. Something like `/pkg/util/httpclient`.

File: modules/overrides/http.go
Comment (joe-elliott) [inline modules/overrides/http.go:30]: we should use the mux to handle this routing

File: modules/overrides/user_config_overrides.go
Comment (joe-elliott) [inline modules/overrides/user_config_overrides.go:227]: this is threadsafe and doe not require locking

File: modules/overrides/user_config_overrides.go
Comment (joe-elliott) [inline modules/overrides/user_config_overrides.go:240]: i find it a little strange (but not wrong) that calling GET to get the limits also updates the local, in memory copy.

i feel like the API should be separate from the mechanism that maintains limits internally. so an api that's just a passthrough to object storage and a separate module that polls tenant limits and provides them as part of the overrides chain.

wdyt? is this separation of concerns worth it?

File: modules/overrides/util.go
Comment (joe-elliott) [inline modules/overrides/util.go:3]: just not the output of this? https://pkg.go.dev/golang.org/x/exp/slices#Contains

File: tempodb/backend/cache/cache.go
Comment (joe-elliott) [inline tempodb/backend/cache/cache.go:114]: we should do this before merging. otherwise it's a difficult to find bug in the future

File: modules/overrides/user_config_limits.go
Comment (joe-elliott) [inline modules/overrides/user_config_limits.go:5]: if this is an api version we should keep it in the path. that's what we do on other endpoints

**Comment by knylander-grafana** [comment]
> > @kvrhdn Will we need doc for this? Should I create a doc issue for it or is it something you'd add to this PR?
> 
> Yes! This is adding a new feature with an API endpoint and it's own set of configurations. Feel free to create a doc issue and assign it to me, I can get a PR started and we can iterate from there 🙂

Thanks, @kvrhdn! The issue is here: https://github.com/grafana/tempo/pull/2543

File: modules/overrides/http.go
Comment (kvrhdn) [reply modules/overrides/http.go:22]: I have addressed this in my last commit by setting the error in `handleError`

File: modules/overrides/http.go
Comment (kvrhdn) [reply modules/overrides/http.go]: Good point, I've split up the handler and register them separately now in modules.go. It's a bit more verbose but not more code overall I think.

File: modules/overrides/util.go
Comment (kvrhdn) [reply modules/overrides/util.go:3]: Woah, today I learned

File: tempodb/backend/cache/cache.go
Comment (kvrhdn) [reply tempodb/backend/cache/cache.go:114]: This comment is misleading, it's intended to say "once we implement caching, we will have to delete stuff from the cache here". We don't have caching yet so I'm just going to delete this comment.

File: modules/overrides/user_config_overrides.go
Comment (kvrhdn) [reply modules/overrides/user_config_overrides.go:227]: I was thinking of a scenario in which someone sends a POST or DELETE request right after we called list. In this case the list we just fetched would be outdated. I think I should take another look at this and simplify it, we really don't need a constant lock for the whole thing...

File: modules/overrides/user_config_overrides.go
Comment (kvrhdn) [reply modules/overrides/user_config_overrides.go:240]: I've linked them with the idea "we have spent work to fetch the data anyway, let's immediately update our local copy". Advantage is you have a quicker update in the instance handling the request. In the context of a distributed system that isn't really a big deal though since all the other components will have to fetch it too...

I'll give separating the two a try and see if it cleans up the code a bit 🙂

File: integration/e2e/overrides_client.go
Comment (joe-elliott) [reply integration/e2e/overrides_client.go:21]: edited the suggested path to `/pkg/httpclient`. overall i'm trying to move things out of util

File: modules/overrides/user_config_overrides.go
Comment (joe-elliott) [reply modules/overrides/user_config_overrides.go:227]: These kinds of state based race conditions are unavoidable. We should just push them to object storage as much as possible.

File: tempodb/backend/cache/cache.go
Comment (joe-elliott) [reply tempodb/backend/cache/cache.go:114]: > To allow merging this PR maybe I can panic instead?

I suppose there's no code path to this b/c you don't implement caching on the overrides api? i'm fine with a panic. heads up @zalegrala

Comment (zalegrala) [reply tempodb/backend/cache/cache.go:114]: Maybe log a message and return nil?  I suppose panic might be warranted if we really never intent to add support.

File: modules/overrides/user_config_overrides.go
Comment (kvrhdn) [reply modules/overrides/user_config_overrides.go]: I've split up the manager, api and client structures. Agree each is simpler now and can focus on its own thing :+1:

So the manager and the API both use client to talk with the backend. All components will run the manager but only the query-frontend needs to run the API.

File: integration/e2e/overrides_client.go
Comment (kvrhdn) [reply integration/e2e/overrides_client.go]: PR to move it: https://github.com/grafana/tempo/pull/2637
Once merged I can update this PR

File: tempodb/backend/cache/cache.go
Comment (kvrhdn) [reply tempodb/backend/cache/cache.go]: I have changed it to return a panic for now. This isn't as clean as returning nil/logging an error but it would ensure we don't accidentally use this in production. If someone isn't careful while testing they might never call delete through a cache or not check the logs of their app. Having the application crash should ensure it doesn't fly under the radar.

File: modules/overrides/user_config_limits.go
Comment (kvrhdn) [reply modules/overrides/user_config_limits.go]: I've opted to drop the version field for now. Usually when we do a breaking change we just introduce a `/v2/` endpoint. Dealing with versioning even before we know how our api could change might be a bit over eager...

File: modules/overrides/user_configurable_overrides.go
Comment (joe-elliott) [inline modules/overrides/user_configurable_overrides.go:182]: i can honestly say i've not yet seen this pattern in go. are you "overriding" the composed interface?

File: modules/overrides/user_configurable_overrides.go
Comment (joe-elliott) [inline modules/overrides/user_configurable_overrides.go:82]: there's a service.TimerService (i'm pretty sure)

File: modules/overrides/user_configurable_overrides.go
Comment (kvrhdn) [reply modules/overrides/user_configurable_overrides.go:182]: It's called struct embedding. This page explains it way better than I ever could: https://eli.thegreenplace.net/2020/embedding-in-go-part-1-structs-in-structs/

By embedding `overrides.Interface` in `userConfigurableOverridesManager` our manager inherits the full implementation. And by overwriting specific methods we can inject our own overrides.
For now we only implemented `Forwarders`, the other methods are automatically handled by the embedded `Interface`.

It's definitely a complicated Go feature but we thought it would be worth it as it saves us a lot of boilerplate and potential of mistakes.

The alternative is to make `overrides.Interface` a regular field, but that would require us to implement each method of `Interface` over again (34 methods!). Every implementation would just look like this:
```go
func (o *userConfigurableOverridesManager) IngestionRateStrategy() string {
	return o.interface.IngestionRateStrategy()
}
```

File: modules/overrides/userconfigurableapi/client.go
Comment (kvrhdn) [reply modules/overrides/userconfigurableapi/client.go:104]: I've added this logic because the local backend would leave empty directories behind. So list returns the tenant even if the overrides file doesn't exist. I'll see if we can address this in the implementation of the local backend.

File: modules/overrides/userconfigurableapi/client.go
Comment (kvrhdn) [reply modules/overrides/userconfigurableapi/client.go:93]: Agree, I think we should log a warning if the exact same path is used. If overrides are stored next to the backend blocks the `overrides/` key will be interpreted as a tenant name and the tenants itself will be parsed as block uuids and fail.

In the documentation I will mention we recommend using 2 buckets but if you use 1 bucket you need to configure a prefix (https://github.com/grafana/tempo/issues/2627#issuecomment-1632729681). Prefixes are supported on all backends so it's a decent solution.

File: modules/overrides/userconfigurableapi/http.go
Comment (kvrhdn) [reply modules/overrides/userconfigurableapi/http.go:96]: Looks like we usually only set status ok and return nothing on successful requests. I'll make this consistent with the other handlers.

File: modules/overrides/userconfigurableapi/http.go
Comment (kvrhdn) [reply modules/overrides/userconfigurableapi/http.go:45]: Changed it so it returns 404 when no overrides are set

File: modules/overrides/userconfigurableapi/http.go
Comment (kvrhdn) [reply modules/overrides/userconfigurableapi/http.go:1]: This file contains the http handlers for `/api/overrides`. `httpclient` is for making request to Tempo?
Do you mean `pkg/api`?

So far we have kept http handlers inside their respective module so it is consistent with the rest of Tempo (afaik). I think we can move out logic if we are doing more than just parsing the request and passing it on to the backend client.

**Review by zalegrala — APPROVED** [review]
This looks good to me.  A few TODOs which we discussed and can follow up in future PRs.

File: modules/overrides/userconfigurableapi/http.go
Comment (joe-elliott) [reply modules/overrides/userconfigurableapi/http.go:1]: so i consider the `/api/overrides` endpoints to be part of Tempo's external api surface which is why I think they should be part of `/pkg/httpclient`

**Review by joe-elliott — APPROVED** [review]
One comment on why I still think we should move the overrides client into `pkg/httpclient`, but I'll leave that up to you. Your call!

**Pull request merged**
